felix/bpf/conntrack/bpf_scanner.go

```diff
@@ -217,18 +235,61 @@ func (s *BPFProgLivenessScanner) RunBPFExpiryProgram(opts ...RunOpt) error {
 	// Record stats...
 	summaryCleanerExecTime.Observe(result.Duration.Seconds())
 
-	gaugeVecConntrackEntries.WithLabelValues("total").Set(float64(
-		cr.NumKVsSeenNormal + cr.NumKVsSeenNATForward + cr.NumKVsSeenNATReverse))
+	total := cr.NumKVsSeenNormal + cr.NumKVsSeenNATForward + cr.NumKVsSeenNATReverse
+
+	gaugeVecConntrackEntries.WithLabelValues("total").Set(float64(total))
 	gaugeVecConntrackEntries.WithLabelValues("normal").Set(float64(cr.NumKVsSeenNormal))
 	gaugeVecConntrackEntries.WithLabelValues("nat_forward").Set(float64(cr.NumKVsSeenNATForward))
 	gaugeVecConntrackEntries.WithLabelValues("nat_reverse").Set(float64(cr.NumKVsSeenNATReverse))
 
-	counterVecConntrackEntriesDeleted.WithLabelValues("total").Add(float64(
-		cr.NumKVsDeletedNormal + cr.NumKVsDeletedNATForward + cr.NumKVsDeletedNATReverse))
+	totalDeleted := cr.NumKVsDeletedNormal + cr.NumKVsDeletedNATForward + cr.NumKVsDeletedNATReverse
+
+	counterVecConntrackEntriesDeleted.WithLabelValues("total").Add(float64(totalDeleted))
 	counterVecConntrackEntriesDeleted.WithLabelValues("normal").Add(float64(cr.NumKVsDeletedNormal))
 	counterVecConntrackEntriesDeleted.WithLabelValues("nat_forward").Add(float64(cr.NumKVsDeletedNATForward))
 	counterVecConntrackEntriesDeleted.WithLabelValues("nat_reverse").Add(float64(cr.NumKVsDeletedNATReverse))
 
+	if !s.scaling {
+		return nil
+	}
+
+	newLiveEntries := int(total - totalDeleted)
+	if s.liveEntries > newLiveEntries {
+		s.higherCount++
+	} else {
+		s.higherCount = 0
+	}
+	s.liveEntries = newLiveEntries
+
+	full := float64(newLiveEntries) / float64(s.maxEntries)
+	log.Debugf("full %f, total %d, totalDeleted %d", full, total, totalDeleted)
+	// If the ct map keeps filling up and gets over 90% full or if it hits 95%
+	// no matter what, resize the map.
+	if s.higherCount >= 3 && full > 0.90 || full > 0.95 {
+		if err := s.writeNewSizeFile(); err != nil {
```

Comment:
Suggest logging a warning here: "Conntrack table nearly full, initiating resize from x to y; will restart to apply new size..."

That'll let the user clearly see what triggered any disruption that they investigate later.